pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -641,6 +646,48 @@ func parseDeviceAddress(addrString string) []string {
 	return addrs
 }
 
+func expandDiskImagesOffline(vmi *v1.VirtualMachineInstance, domain *api.Domain) {
+	logger := log.Log.Object(vmi)
+	for _, disk := range domain.Spec.Devices.Disks {
+		if shouldExpand(disk) {
+			err := expandDiskImageOffline(disk.Source.File, disk.Size)
+			if err != nil {
+				logger.Reason(err).Errorf("failed to expand disk image %v at boot", disk)
+			}
+		}
+	}
+}
+
+func shouldExpand(disk api.Disk) bool {
+	if disk.AlwaysExpand {
+		return true
+	}
+	diskInfo, err := converter.GetImageInfo(getSourceFile(disk))
+	if err != nil {
+		return false
+	}
+	if diskInfo.VirtualSize >= disk.Size {
+		return false
+	}
+	return true
+}
+
+func expandDiskImageOffline(imagePath string, size int64) error {
+	log.Log.Infof("pre-start expansion of image %s to size %d", imagePath, size)
+	var preallocateFlag string
+	if converter.IsPreAllocated(imagePath) {
+		preallocateFlag = "--preallocation=falloc"
+	} else {
+		preallocateFlag = "--preallocation=off"
+	}
+	cmd := exec.Command("/usr/bin/qemu-img", "resize", preallocateFlag, imagePath, strconv.FormatInt(size, 10))
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("Expanding image failed with error: %v, output: %s", err, out)
+	}
```

Comment:
if the overhead calculation for the FS is incorrect (trying to expand disk too large for PVC), will this result in the VMI never booting?

Reply:
If we are making the disk bigger, there's two scenarios. Preallocated, where fallocate will fail and not modify the disk. and not-preallocated (sparse) where we might run out of space later on as the disk is filled out with actual data. Both of those scenarios do not result in an unbootable VMI.

If we are making the disk smaller, yes, that is a risky operation that might corrupt data. In newer qemu-img than is available on kubevirt it explicitly fails without a --shrink flag. I test for this scenario in the shouldExpand function to avoid it happening.